### PR TITLE
Deprecate jsxBracketSameLine option in favour of multi-language bracketSameLine option

### PR DESCRIPTION
**Pull request by kurtztech**
## Description

I've added a new option `bracketSameLine` that was approved to address issue #5377, essentially making a generic version of the `jsxBracketSameLine` option that works for HTML, JSX, Vue, and Angular. I've deprecated `jsxBracketSameLine`. It still works, but it is now listed as deprecated in the docs and CLI, and has been removed from the playground options.

## Checklist

- [x] I’ve added tests to confirm my change works.
- [x] (If changing the API or CLI) I’ve documented the changes I’ve made (in the `docs/` directory).
- [x] (If the change is user-facing) I’ve added my changes to `changelog_unreleased/*/XXXX.md` file following `changelog_unreleased/TEMPLATE.md`.
- [x] I’ve read the [contributing guidelines](https://github.com/prettier/prettier/blob/main/CONTRIBUTING.md).

**✨[Try the playground for this PR](https://prettier.io/playground-redirect)✨**

### CODE REVIEW

**Review by alexander-akait — COMMENTED** [review]
We should keep old options (include CLI) for compatibility, just log deprecation

File: tests/format/markdown/markdown/real-world-case.md [hunk 1]
```diff
@@ -25,50 +25,70 @@ conforms to a consistent style. (See this [blog post](http://jlongster.com/A-Pre
...
-  + [Visual Studio Code](#visual-studio-code)
-  + [Visual Studio](#visual-studio)
-  + [Sublime Text](#sublime-text)
-  + [JetBrains WebStorm, PHPStorm, PyCharm...](#jetbrains-webstorm-phpstorm-pycharm)
-* [Language Support](#language-support)
-* [Related Projects](#related-projects)
-* [Technical Details](#technical-details)
-* [Badge](#badge)
-* [Contributing](#contributing)
+- [Prettier](#prettier)
```
Comment (fisker) [inline tests/format/markdown/markdown/real-world-case.md:28]: Can you revert change in this file?

File: tests/format/jsx/comments/jsfmt.spec.js [hunk 2]
```diff
@@ -1,3 +1,10 @@
+run_spec(__dirname, ["flow", "babel", "typescript"], {
+  bracketSameLine: true,
+  errors: {
+    typescript: ["in-end-tag.js"],
+  },
+});
```
Comment (fisker) [inline tests/format/jsx/comments/jsfmt.spec.js:6]: I don't think we need new tests for jsx, replace all `jsxBracketSameLine` with new `bracketSameLine`.

We only need one simple test to make sure `jsxBracketSameLine` still work for `jsx`.

File: tests/format/vue/vue/__snapshots__/jsfmt.spec.js.snap [hunk 3]
```diff
@@ -1594,6 +2123,25 @@ export default {
...
+<template>
+	<Input></Input>
+</template>
+
+=====================================output=====================================
+<template>
+  <Input></Input>
+</template>
+
+================================================================================
```
Comment (fisker) [inline tests/format/vue/vue/__snapshots__/jsfmt.spec.js.snap:2142]: This file is hard to review, can you add a new test?

File: tests/format/markdown/markdown/real-world-case.md (hunk 1 above)
Comment (kurtztech) [reply tests/format/markdown/markdown/real-world-case.md:28]: I now see this comment just above, but there doesn't seem to be a `yarn toc` script:

```
<!-- Do not edit TOC, regenerate with `yarn toc` -->
```

File: tests/format/vue/vue/__snapshots__/jsfmt.spec.js.snap (hunk 3 above)
Comment (kurtztech) [reply tests/format/vue/vue/__snapshots__/jsfmt.spec.js.snap]: created new tests

**Comment by kurtztech** [comment]
@fisker is there anything else I need to do before this PR can be accepted?

**Comment by fisker** [comment]
I reverted unnecessary changes, and wrote some new tests.

File: changelog_unreleased/html/11006.md [hunk 4]
```diff
@@ -0,0 +1,34 @@
+#### Replace jsxBracketSameLine option with bracketSameLine option (#11006 by @kurtztech)
+
+<!-- deprecate the jsxBracketSameLine option in favour of an bracketSameLine option that will work for -->
+<!-- HTML, Angular, Vue, JSX  -->
```
Comment (fisker) [inline changelog_unreleased/html/11006.md:4]: This should not in comments

File: src/language-js/options.js [hunk 5]
```diff
@@ -32,7 +32,8 @@ module.exports = {
     category: CATEGORY_JAVASCRIPT,
     type: "boolean",
     default: false,
-    description: "Put > on the last line instead of at a new line.",
+    description:
```
Comment (fisker) [inline src/language-js/options.js:35]: Is it possible to use `deprecated` option? https://github.com/ikatyang/vnopts#deprecated

So maybe we can warn about `--jsx-bracket-same-line` on CLI `node bin/prettier index.js --jsx-bracket-same-line -l`

**Review by fisker — CHANGES_REQUESTED** [review]

File: changelog_unreleased/html/11006.md [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+  id="foo-bar-baz"
+  class="bar-foo-baz"
+  title="a sample title"
+  data-foo="bar"
+  data-bar="baz"
+  data-baz="foo">
+  lorem ipsum dolor sit amet
+</div>
+
+
```
Comment (fisker) [inline changelog_unreleased/html/11006.md:22]:
```suggestion

````

File: src/language-html/options.js [hunk 7]
```diff
@@ -32,4 +32,11 @@ module.exports = {
     default: false,
     description: "Indent script and style tags in Vue files.",
   },
+  bracketSameLine: {
```
Comment (fisker) [inline src/language-html/options.js:35]: I think this should not in `html` but as a common option? I'm not sure how option category works, but seems weird that this option in HTML category, but effect `JSX`.

Comment (kurtztech) [reply src/language-html/options.js:35]: I've moved it to the common options

File: tests/format/markdown/markdown/real-world-case.md (hunk 1 above)
Comment (fisker) [reply tests/format/markdown/markdown/real-world-case.md]: This is only a test file, content don't need to be the same as the docs.

**Comment by fisker** [comment]
Can you fix tests?

**Comment by kurtztech** [comment]
@fisker tests are fixed, I just had to removed the `default` property from the `jsxBracketSameLine` option after deprecation.

File: src/language-js/options.js [hunk 8]
```diff
@@ -26,13 +26,14 @@ module.exports = {
...
   },
+  bracketSameLine: commonOptions.bracketSameLine,
   bracketSpacing: commonOptions.bracketSpacing,
   jsxBracketSameLine: {
     since: "0.17.0",
     category: CATEGORY_JAVASCRIPT,
     type: "boolean",
-    default: false,
     description: "Put > on the last line instead of at a new line.",
+    deprecated: "2.4.0",
```
Comment (fisker) [inline src/language-js/options.js:36]: We had test for deprecated options before, can we add one for this?  Check https://github.com/prettier/prettier/pull/7536/files, the old `tests_integration/__tests__/arg-parsing.js` file is moved to `tests/integration/__tests__/arg-parsing.js`

**Review by fisker — APPROVED** [review]
LGTM.

One comment about test.

I'm not good at docs, we'll need other maintainers to approve.

**Comment by okmaon** [comment]
exactly what i was looking for! would love to use this feature, looking forward for the release

kudos kurtztech ✨

**Comment by xianshenglu** [comment]
Looking forward for the release...

**Comment by gethari** [comment]
Waiting for this !

**Review by sosukesuzuki — COMMENTED** [review]
This PR should not include updating markdown files under `/website/versioned_docs/version-stable/`. It should be updated when we release 2.4.0.

**Review by alexander-akait — APPROVED** [review]
Let's fix problems above and I think we can merge it

**Comment by kurtztech** [comment]
> Let's fix problems above and I think we can merge it

So is it just the markdown file changes I need to remove, or are there other issues?

**Comment by sosukesuzuki** [comment]
@kurtztech Please check a comment from @fisker (https://github.com/prettier/prettier/pull/11006#discussion_r655119897)

**Comment by gethari** [comment]
Looking forward for the release.

**Comment by kieranbarlow** [comment]
Do we have a date for when this will be available?

**Comment by sosukesuzuki** [comment]
> I have time to work on it this week finally.

@kurtztech Ooops!! Sorry! I missed your message. I've pushed commits...

**Review by fisker — APPROVED** [review]

**Review by sosukesuzuki — APPROVED** [review]

**Comment by sosukesuzuki** [comment]
We'll merge this when we have a schedule for the next minor release.

**Review by alexander-akait — APPROVED** [review]

**Comment by r0unflame** [comment]
When are you going to release this?

**Comment by sosukesuzuki** [comment]
next release schedule is undecided yet.

**Comment by jasperf** [comment]
Would love to see this merged and have brackets display in a more sensible way.

**Comment by HADB** [comment]
Great job! 👍🏻

**Comment by glen-84** [comment]
The option name is super generic ... https://github.com/prettier/prettier/issues/5377#issuecomment-689393993.

**Pull request merged**